### ISSUE TITLE
How to implement reconnect logic when peer uses same ID? 

### ISSUE THREAD

**Issue by jeffRTC** [paullouisageneau/libdatachannel#516]
@paullouisageneau 

I have following logic as per now but it doesn't work as expected when client uses same ID over again. 

```
        pc->onStateChange([&, id](rtc::PeerConnection::State state) {
            std::cout << "State: " << state << "\n";

            if(state == rtc::PeerConnection::State::Closed) {
                std::scoped_lock guard{dataChannelMapMutex, peerConnectionMapMutex, localCandidatesMapMutex};
                
                dataChannelMap.erase(id);

                peerConnectionMap.erase(id);

                localCandidatesMap.erase(id);
            }
        });
```

It works perfectly if I omit all unordered map erases, but that brings another problem which is dataChannelMap having a channel that's maybe open or closed much like schrodinger's cat.

**Comment by jeffRTC** [comment]
@paullouisageneau 

I did more debugging on this issue. It appear that if you refresh the page, onStateChange with `rtc::PeerConnection::State::Closed` never get called on sever side unless you start a new connection. 

I expect the correct behavior to be onState should get called as soon as you refresh the page. 

Once a new connection made with same id,

onStateChange get called twice with `rtc::PeerConnection::State::Closed` 

This doesn't happens if,

1 ) I use a new id
2 ) Restart server application
3 )  Omit all unordered map erases.

**Comment by jeffRTC** [comment]
@paullouisageneau 

I even added slight 20 seconds delay on client application before starting a new connection to see if onStateChange get invoked in Server side, but nothing happens.

**Comment by paullouisageneau** [comment]
I think there is a misunderstanding about `rtc::PeerConnection::State`: if the connection is closed by the remote side or lost, the state will change to `Disconnected` or `Failed` (If the remote peer disappears abruptly, the connection will take a while to timeout, like 30s or so, depending on RTT), but not `Closed`. The state will change to `Closed` only when you call `close()` on the `PeerConnection` or when you destroy it.

To prevent callback mismatch when a new `PeerConnection` is created for the same peer ID, you should first call `close()` on the old `DataChannel` and `PeerConnection`. Since it is guaranteed no callbacks will be called after `close()` returns, you can create a new `PeerConnection` safely.

**Comment by jeffRTC** [comment]
@paullouisageneau 

But the problem is that you can't really close old Data Channel from browser when you refresh the page.

**Comment by paullouisageneau** [comment]
I meant from the server. When the server receives a request to create a new connection for an existing peer ID, it needs to close the old one first.

**Comment by jeffRTC** [comment]
@paullouisageneau Thanks works but again seems can't use locks when removing from map.  Deadlocks..

I need time to think and summarize, will open new issue for that.
